spec/factories/openid_users.rb

```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :openid_user, class: 'OpenidUser' do
+    transient do
+      identity_attrs nil
+    end
+
+    callback(:after_build, :after_stub) do |user, t|
+      if t.identity_attrs
+        # Normally the identity attribute would be set by looking up a previously-saved identity.
+        # This is a conventience method for allowing the identity to be constructed and associated
+        # with the user without having to save it.
+        user.instance_variable_set(:@identity, OpenidUserIdentity.new(t.identity_attrs))
+      end
```

Comment:
Extracting attributes for the user identity classes into the `user_identity_attrs` factory, and then transforming test setup code like this makes sense to me:

```
- let(:user) { build(:user, :loa3) }
+ let(:user) { build(:user, identity_attrs: build(:user_identity_attrs, :loa3)) }
```

It's a little more verbose but it's also more explicit. When I first looked at the tests I wondered how the tests were using a `User` object without first creating a `UserIdentity` object as I had seen multiple places in the application code. The alternative seems to be duplicating a lot of test user details across the `user` and `openid_user` factories. I'm curious what other folks think before I go ahead with those changes (~115).

Reply:
We didn't have a concept of UserIdentity initially and there were a LOT of tests that would have broken if I didn't introduce the "transient" methods on the User factory for building UserIdentity dynamically. I think it's convenient to do it that way for now -- but one longer term better approach would be to build a SamlAssertion in a helper, which then builds User and UserIdentity from the SAMLAssertion and MviProfile factories.